npm/podController.go

```diff
@@ -0,0 +1,634 @@
+
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer c.workqueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			// As the item in the workqueue is actually invalid, we call
+			// Forget here else we'd go into a loop of attempting to
+			// process a work item that is invalid.
+			c.workqueue.Forget(obj)
+			utilruntime.HandleError(fmt.Errorf("expected string in workqueue but got %#v", obj))
+			return nil
+		}
+		// Run the syncHandler, passing it the namespace/name string of the
+		// Pod resource to be synced.
+		// TODO : may consider using "c.queue.AddAfter(key, *requeueAfter)" according to error type later
+		if err := c.syncPod(key); err != nil {
+			// Put the item back on the workqueue to handle any transient errors.
+			c.workqueue.AddRateLimited(key)
+			return fmt.Errorf("error syncing '%s': %s, requeuing", key, err.Error())
+		}
+		// Finally, if no error occurs we Forget this item so it does not
+		// get queued again until another change happens.
+		c.workqueue.Forget(obj)
+		log.Logf("Successfully synced '%s'", key)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// syncPod compares the actual state with the desired, and attempts to converge the two.
+func (c *podController) syncPod(key string) error {
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	// Get the Pod resource with this namespace/name
+	pod, err := c.podLister.Pods(namespace).Get(name)
+	// TODO: Reduce scope of lock later
+	c.npMgr.Lock()
+	defer c.npMgr.Unlock()
+	if err != nil {
+		if errors.IsNotFound(err) {
+			fmt.Printf("pod %s not found, may be it is deleted\n", key)
+			// Find the pod object from a local cache and start cleaning up process (calling cleanUpDeletedPod function)
+			cachedPod, found := c.npMgr.PodMap[key]
+			if found {
+				// TODO: better to use cachedPod when calling cleanUpDeletedPod?
+				err = c.cleanUpDeletedPod(cachedPod.getPodObjFromNpmObj())
+				if err != nil {
+					metrics.SendErrorLogAndMetric(util.PodID, "Error: cannot delete ipset due to %v when pod is not found", err)
+					// cleaning process was failed, need to requeue and retry later.
+					return fmt.Errorf("Cannot delete ipset due to %s when pod is not found\n", err.Error())
+				}
+			}
+		}
+		return err
+	}
+
+	if pod.DeletionTimestamp != nil || pod.DeletionGracePeriodSeconds != nil {
+		err = c.cleanUpDeletedPod(pod)
+		if err != nil {
+			metrics.SendErrorLogAndMetric(util.PodID, "Error: cannot delete ipset due to %v when DeletionTimestamp or DeletionGracePeriodSeconds is not nil", err)
+			return fmt.Errorf("Failed to clean up pod due to %s when DeletionTimestamp or DeletionGracePeriodSeconds is not nil\n", err.Error())
+		}
```

Comment:
We will need to return here, or else we will be cleaning up and adding again

Reply:
applied